src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
 
             void IDisposable.Dispose()
             {
-                this.Presenter.AssertIsForeground();
+            }
 
-                // VS is letting go of us.  i.e. because a new FAR call is happening, or because
-                // of some other event (like the solution being closed).  Remove us from the set
-                // of sources for the window so that the existing data is cleared out.
-                Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 1);
-                Debug.Assert(_findReferencesWindow.Manager.Sources[0] == this);
+            #endregion
 
-                _findReferencesWindow.Manager.RemoveSource(this);
+            private void RemoveSink(ITableDataSink sink)
+            {
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Remove(arg), sink);
```

Comment:
please don't use Interlocked here. We have UI thread affinitized data that should just use normal constructs, and we have free-threaded data that should be protected by _gate.  This mixes concepts in terms of how data should be operated upon.

Reply:
Yes, I agree. In fact as I was writing my own source table code, I found that forcing sequential handling of events was necessary to make sure my data integrity would be resilient to concurrent execution, etc.
That said, I'm not proposing any breaking changes here. AFAIK the table manager itself isn't changing so no old code should be broken. 
Assuming the scenario owner (Roslyn's feature at this point) keeps the table control / manager instance concealed from the outside world, they continue to be fine. But to support multiple sinks, it is likely that the table manager instance will have to be shared, which allows folks to enumerate the various sources and add their own sinks to them. So in preparation for Roslyn exposing the table manager (which would be a subsequent PR), we first need to prepare Roslyn for dealing with multiple sinks to avoid bugs. That's what this PR is about.

Reply:
Sure.  But in this case, i would just keep the threading assert (first in the method), and just do a normal .Add or .Remove on the immutable-array/list.  The Interlocked serves no purpose.  Either we're UI thread only, in which case it's unnecessary.  Or we're free-threaded, in which case none of the code is corrrect, and being Interlocked doesn't change that.  Thanks! :)

Reply:
Fair enough. I'm removing the interlocks and asserting the ui thread everywhere.

Reply:
Thanks! :)